### PR TITLE
Enhance README organization, add more background info to code

### PR DESCRIPTION
**Pull request by GavinRay97**
Heya,

When I first read the `Rune` readme, there were a couple of things I thought could be improved:
- I didn't understand where methods like `appendMotheredHuman` etc come from. We "magically" have access to something we didn't define and it feels strange. It gets described later on, but it would have been nice for a comment to say "Hey, don't worry! I'll explain why this magic-method exists in a little bit!"
- DataDraw doesn't get mentioned, and I had to piece the connection together myself. In my opinion, a HUGE amount of appeal comes from the fact that DD is integrated into Rune, and it's very interesting + explains much of the behavior and design decisions/architecture + performance.

Also, many people will want to jump directly to specific things ("how do I build this?", "show me a code example?") without reading the rest of it, so I tried to add enough context to each section that it would make sense if you went straight there and skipped the rest.

Lastly, I added some organizational structure and a TOC so folks can navigate the page easier.

Hope this is helpful, I'd like to highlight Rune in the best way possible for potential users as I think it has some neat ideas 🙂 

---

Ahh, one last thing -- I chose the languages for the syntax highlighting for each code block based on which one looked the best/could highlight the majority of the tokens:

Go:
```go
// Check the MAC (message authentication code) for a message.  A MAC is derived
func checkMac(macSecret: secret(string), message: string, mac: string) -> bool {
    computedMac = computeMac(macSecret, message)
    return mac == computedMac
}
```

Rust (TypeScript comes close too, but not quite as good):
```rs
class Human(self: Human, name: string, mother: Human? = null(self), father: Human? = null(self)) {
  self.name = name
  if !isnull(mother) {
    mother.appendMotheredHuman(self)
  }
}
```

### CODE REVIEW

**Comment by GavinRay97** [comment]
Sorry about the random formatting changes that removed only spaces in the diff.
VS Code auto-format did that -- I didn't catch it in the PR. I can resubmit if it's an issue.

**Comment by GavinRay97** [comment]
I just learned about the `dataview` tool and figured out what it does, it'd be neat to mention that as well in the README!

image

File: README.md [hunk 1]
```diff
@@ -7,21 +8,46 @@ This is not an officially supported Google product.
...
+1. A simple example of a secret type, and how it can be used to protect against timing attacks.
+2. A more complex example of a data-structure with nullable, self-referential recursive references.
+    - _(This can present footguns in many languages, but it is trivial in Rune)_
+
+## Dedicated "secret" types
 
 Consider the following example for treatment of secrets:
 
-```
+```go
```
Comment (0dminnimda) [inline README.md:50]: Although it's nice to see the syntax highlighting we probably should use `rune` as it's in the other documents, like [rune4python.md](https://github.com/google/rune/blob/92f09e0f46b56d988b447bacd977e04ea76aba45/g3doc/rune4python.md?plain=1#L43)

It should be nice to eventually submit the syntax to [linguist](https://github.com/github/linguist/blob/master/vendor/README.md) which github uses for syntax highlighting, it look to be possible
Also there's was a request to allow github to support custom languages https://github.com/github/linguist/issues/2627

Comment (GavinRay97) [reply README.md:50]: Ah, is there a TextMate grammar for Rune syntax that could be submitted to Linguist for the highlighting?
I had looked for one to try to add it into VS Code for proper highlighting but didn't notice one in the repo so far

Agreed long-term that's definitely the best solution

Comment (0dminnimda) [reply README.md:50]: Right now Rune uses yacc+bison for parsing, you can look into [parse/deparse.y](https://github.com/google/rune/blob/main/parse/deparse.y) also there's grammar for vim [editors/vim/syntax/rune.vim](https://github.com/google/rune/blob/main/editors/vim/syntax/rune.vim) and that's it.
TextMate grammar needs to be created from scratch.

But the language is still young and grammar can change anytime.
I don't think it's worth the effort right now.

**Comment by waywardgeek** [comment]
Thanks for all the work on documentation!

**Pull request merged**